assembly/src/parsers/io_ops.rs

```diff
@@ -207,32 +263,84 @@ pub fn parse_storew(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), Ass
 /// when immediate value is 1, `PUSH` operation is replaced with `PAD INCR` because in most cases
 /// this will be more efficient than doing a `PUSH`.
 ///
 /// # Errors
 ///
-/// This function expects an assembly op with exactly one immediate value that is a valid field
-/// element in decimal or hexadecimal representation. It will return an error if the immediate
-/// value is invalid or missing. It will also return an error if the op token is malformed or
-/// doesn't match the expected instruction.
+/// It will return an error if no immediate value is provided or if any of parameter formats are
+/// invalid. It will also return an error if the op token is malformed or doesn't match the expected
+/// instruction.
 fn parse_push_constant(span_ops: &mut Vec<Operation>, op: &Token) -> Result<(), AssemblyError> {
-    // validate op
-    validate_op_len(op, 1, 1, 1)?;
-    if op.parts()[0] != "push" {
-        return Err(AssemblyError::unexpected_token(op, "push.{param}"));
+    let param_idx = 1;
+    validate_op_len(op, param_idx, 1, MAX_CONST_INPUTS)?;
+
+    let param_count = op.num_parts() - param_idx;
+    // for multiple input parameters, parse & push each one onto the stack in order, then return
+    if param_count > 1 {
+        for param_idx in param_idx..=param_count {
+            let value = parse_element_param(op, param_idx)?;
+            push_value(span_ops, value);
+        }
+        return Ok(());
     }
 
-    // update the span block
-    let value = parse_element_param(op, 1)?;
-    push_value(span_ops, value);
+    // for a single input, there could be one value or there could be a series of many hexadecimal
+    // values without separators
+    let param_str = op.parts()[param_idx];
+    if let Some(param_str) = param_str.strip_prefix("0x") {
+        // parse 1 or more hexadecimal values
+        let values = parse_hex_params(op, param_idx, param_str)?;
+        // push each value onto the stack in order
+        for &value in values.iter() {
+            push_value(span_ops, value);
+        }
+    } else {
+        // parse 1 decimal value and push it onto the stack
+        let value = parse_decimal_param(op, param_idx, param_str)?;
+        push_value(span_ops, value);
+    }
 
     Ok(())
 }
 
+/// Parses a hexadecimal string into a vector of 1 or more field elements, up to 16 total. If more
+/// than one value is specified, then the total number of specified bytes must be a multiple of 8.
+fn parse_hex_params(
+    op: &Token,
+    param_idx: usize,
+    param_str: &str,
+) -> Result<Vec<BaseElement>, AssemblyError> {
+    let is_single_element = param_str.len() <= HEX_CHUNK_SIZE;
+    let is_invalid_param =
+        param_str.len() % 2 != 0 || !is_single_element && param_str.len() % HEX_CHUNK_SIZE != 0;
+
+    if is_invalid_param {
+        Err(AssemblyError::invalid_param(op, param_idx))
+    } else if param_str.len() > HEX_CHUNK_SIZE * MAX_CONST_INPUTS {
+        Err(AssemblyError::extra_param(op))
```

Comment:
Took me a while to parse through this. Maybe we should improve readability here a bit (even at the expense of a few extra lines)? Maybe something like:
```Rust
let is_single_element = if param_str.len() <= HEX_CHUNK_SIZE {
    if param_str.len() % 2 != 0 {
        return Err(AssemblyError::invalid_param(op, param_idx));
    }
    true
} else {
    if param_str.len() % HEX_CHUNK_SIZE != 0 {
        return Err(AssemblyError::invalid_param(op, param_idx));
    } else if param_str.len() > HEX_CHUNK_SIZE * MAX_CONST_INPUTS {
        return Err(AssemblyError::extra_param(op));
    }
    false
};
```
And then we can handle non-error cases after this.

Reply:
Yeah, this makes sense.